### ISSUE TITLE
Question: [uwp] Is Canvas optimized for huge sizes (1 million px width or height)?

### ISSUE THREAD

**Issue by jtorjo** [microsoft/microsoft-ui-xaml#3468]
I have a `Canvas` control with around 300 children controls, which get resized as the user increases/decreases a zoom factor.
Worst case scenario, this canvas would end up having lets say 1,000,000 pixels in width.

My assumption is that the UI layer would simply figure out what needs to be drawn from each of the canvas's children and draw only that -- so it would be very efficient. 

Is my assumption correct, or do I need to actually implement drawing myself?

### Background

I'm implementing a video editor, and what I'm talking about here is the timeline. If the user has, lets say 30 minutes of footage, and zooms in a lot, lets say 1 second = 500px, we end up at 900,000 px)

**Comment by stevewri** [comment]
XAML specific question about the canvas control.

**Comment by ranjeshj** [comment]
@jtorjo If the content is that big and has 300 children controls, perhaps some sort of virtualization is in order to get decent perf.

**Comment by jtorjo** [comment]
@ranjeshj  Thanks. I do get some virtualization would be needed - but how do you actually implement that? I don't see a `VirtualCanvas` class. 

One idea I have is to hide the controls that are not visible - is this what you mean?

**Comment by ranjeshj** [comment]
Right, you would have to look at the Scrollviewer's viewport and just measure/arrange items within what. Another option could be to see if you can use ItemsRepeater and write a custom virtualizing layout (where you get the rect of what is visible/realized). That way your custom layout can be de-coupled from the ScrollViewer.

**Comment by jtorjo** [comment]
@ranjeshj I ended up doing something different altogether. Basically, I would have a `VirtualPosition`, and a  `VirtualWidth` for the canvas and its children. The Canvas itself its regular size, and I've implemented another `Scroll` control that would operate on `VirtualPosition` and `VirtualWidth`. Thus, I would constantly move/show/hide all children based on their `VirtualPosition` and `VirtualWidth`. This is insanely fast, and `VirtualWidth` can go into the millions of pixels
